### PR TITLE
[Merged by Bors] - feat: port Analysis.Calculus.ContDiff

### PR DESCRIPTION
**Pull request by urkud**
---

- [x] depends on: #4249
- [x] depends on: #4256
- [x] depends on: #4542
- [x] depends on: #4564

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#4249
* leanprover-community/mathlib4#4256
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by fpvandoorn** [comment]
I hope you don't mind if I add some commits here

**Comment by urkud** [comment]
@fpvandoorn I had non-pushed work

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#4249~~
* ~~leanprover-community/mathlib4#4256~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#4249~~
* ~~leanprover-community/mathlib4#4256~~
* ~~leanprover-community/mathlib4#4542~~
* ~~leanprover-community/mathlib4#4564~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by jcommelin** [comment]
Is the style linting the only thing that is currently failing here?

**Comment by urkud** [comment]
@jcommelin I'll re-enable `maxHeartbeats` on 1 lemma and add some porting comments tonight.

File: Mathlib/Analysis/Calculus/ContDiff.lean [hunk 1]
```diff
@@ -2517,202 +2368,238 @@ theorem ContinuousLinearMap.norm_iteratedFderivWithin_le_of_bilinear (B : E →L
...
   exact
     ContinuousLinearMap.compL 𝕜 Eu (Fu →L[𝕜] G) (Fu →L[𝕜] Gu)
       (ContinuousLinearMap.compL 𝕜 Fu G Gu (isoG.symm : G →L[𝕜] Gu)) Bu₀
+  have hBu : Bu = ContinuousLinearMap.compL 𝕜 Eu (Fu →L[𝕜] G) (Fu →L[𝕜] Gu)
+      (ContinuousLinearMap.compL 𝕜 Fu G Gu (isoG.symm : G →L[𝕜] Gu)) Bu₀ := rfl
   have Bu_eq : (fun y => Bu (fu y) (gu y)) = isoG.symm ∘ (fun y => B (f y) (g y)) ∘ isoD := by
     ext1 y
-    simp only [Bu, ContinuousLinearMap.compL_apply, Function.comp_apply,
-      ContinuousLinearMap.coe_comp', LinearIsometryEquiv.coe_coe'', ContinuousLinearMap.flip_apply,
-      LinearIsometryEquiv.apply_symm_apply]
```
Comment (hrmacbeth) [inline Mathlib/Analysis/Calculus/ContDiff.lean]: Include the deleted
```lean
    simp only [Bu, ContinuousLinearMap.compL_apply, Function.comp_apply,
      ContinuousLinearMap.coe_comp', LinearIsometryEquiv.coe_coe'', ContinuousLinearMap.flip_apply,
      LinearIsometryEquiv.apply_symm_apply]
```
in a porting note?

File: Mathlib/Analysis/Calculus/ContDiff.lean [hunk 2]
```diff
@@ -0,0 +1,2817 @@
...
+    [NormedAddCommGroup Du] [NormedSpace 𝕜 Du] [NormedAddCommGroup Eu] [NormedSpace 𝕜 Eu]
+    [NormedAddCommGroup Fu] [NormedSpace 𝕜 Fu] [NormedAddCommGroup Gu] [NormedSpace 𝕜 Gu]
+    (B : Eu →L[𝕜] Fu →L[𝕜] Gu) {f : Du → Eu} {g : Du → Fu} {n : ℕ} {s : Set Du} {x : Du}
+    (hf : ContDiffOn 𝕜 n f s) (hg : ContDiffOn 𝕜 n g s) (hs : UniqueDiffOn 𝕜 s) (hx : x ∈ s) :
+    ‖iteratedFDerivWithin 𝕜 n (fun y => B (f y) (g y)) s x‖ ≤
+      ‖B‖ * ∑ i in Finset.range (n + 1), (n.choose i : ℝ) * ‖iteratedFDerivWithin 𝕜 i f s x‖ *
+        ‖iteratedFDerivWithin 𝕜 (n - i) g s x‖ := by
+  /- We argue by induction on `n`. The bound is trivial for `n = 0`. For `n + 1`, we write
+    the `(n+1)`-th derivative as the `n`-th derivative of the derivative `B f g' + B f' g`, and apply
+    the inductive assumption to each of those two terms. For this induction to make sense,
```
Comment (urkud) [inline Mathlib/Analysis/Calculus/ContDiff.lean:2253]:
```suggestion
    the `(n+1)`-th derivative as the `n`-th derivative of the derivative `B f g' + B f' g`, and
    apply the inductive assumption to each of those two terms. For this induction to make sense,
```

File: Mathlib/Analysis/Calculus/ContDiff.lean (hunk 1 above)
Comment (Parcly-Taxel) [reply Mathlib/Analysis/Calculus/ContDiff.lean]: Done.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**